### ISSUE TITLE
Avoid storing a list of outputs to compute aggregated metrics at the end of the epoch.

### ISSUE THREAD

**Issue by rotabulo** [PyTorchLightning/pytorch-lightning#3728]
Hi, 
  to my understanding, the current way of logging an aggregated metric at the end of an epoch requires implicitly to store the outputs of all steps in the epoch. There are tasks like semantic segmentation requiring to accumulate a confusion matrix over steps, e.g. in order to compute the mIoU metric. However, storing all those confusion matrices for the only sake of aggregating them at the end is a waste of resources (it might occupy gigabytes of memory). Do you have a preferred solution to sidestep the issue? It would be better to also offer the possibility of aggregating information while traversing the steps (without storing all outputs) and let the aggregated data to be accessible at the end of the epoch e.g. for logging.

Thank you.

**Comment by awaelchli** [comment]
the metrics guys are currently working on this :)

**Comment by SkafteNicki** [comment]
@rotabulo sorry for first getting back to you now.
The new metric API support states, which should solve your problems. Take accuracy as an example:
https://github.com/PyTorchLightning/pytorch-lightning/blob/3abfec896212ea85e45d6ac3ccb323ef242d16de/pytorch_lightning/metrics/classification/accuracy.py#L97-L115
In the update state we simply keep adding the number of correct matches and the total number of samples (thus we only store 2 numbers for this metric, regardless of the number of batches). The final accuracy is computed in the `compute` method.
For your use case, you could do the same, having a `self.confmat` as a state which you just keep adding to in `update` method. In this case you would only have to store `num_classes x num_classes`. 
Please see the documentation for more info about the new interface: https://pytorch-lightning.readthedocs.io/en/latest/metrics.html

**Comment by rotabulo** [comment]
@SkafteNicki I figured that actually out already when I saw the commit. I updated my code for the mIoU computation using the new API already and it works great. Thanks

**Comment by SkafteNicki** [comment]
@rotabulo sound good, happy that the new API works for you :] 
Closing this issue.

**Comment by rotabulo** [comment]
@SkafteNicki I have a concern about the current implementation. I haven't time to dig in depth into the code, but I saw that the metric states are stored as buffers and when I use DistributedDataParallel of Pytorch, at the beginning of each forward pass, the buffers of device 0 are replacing the ones of the other devices. Hence, if the metric object keeps track of the metrics per device without syncronizing at each step, it will loose information and not compute the metric correctly at the end of the epoch. Is this potential issue taken care of in case the per-step computation is not enabled in the metric object?

**Comment by SkafteNicki** [comment]
@rotabulo I am not completely sure if I understand your concern. Can you provide some more details/code?
We have tests in place that should secure that the correct value is calculated, with or without the `dist_sync_on_step` flag enabled.
That said, the API is pretty new so there may be something that we have missed :]

**Comment by rotabulo** [comment]
@SkafteNicki after some preliminary checks of my concern, it looks like this might be indeed an issue. I try to give more details of the issue. DistributedDataParallel of Pytorch synchronizes all buffers within a model at the beginning of each forward pass by copying the one that is in the device with rank 0 to the rest of the devices of a node. Since the new Metrics class exploits model buffers to store states, it is subject to the same behavior, i.e. at the beginning of the forward pass, the metric states of device 0 will overwrite the metric states of the other devices. This will induce a loss of information and therefore the metric will result in an incorrect value at the end of the epoch..

Do you have a unit test where the model is wrapped by DistributedDataParallel (either directly or via Trainer)?

Here is a code snipped that reproduces the issue:
```
import torch
import time

import pytorch_lightning as pl

class SumMetric(pl.metrics.Metric):
    def __init__(self):
        super().__init__(dist_sync_on_step=False)
        self.add_state('sum',default=torch.tensor(0.), dist_reduce_fx='sum')

    def update(self, preds, target=None):
        self.sum+=preds.sum()

    def compute(self):
        return self.sum

data=torch.arange(10)[:,None].float()

dataset = torch.utils.data.TensorDataset(data)


class IdentityModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.metric=SumMetric()
        # this is here just to have some fake parameters in the model
        self.params=torch.nn.Conv2d(10,10,1)

    def forward(self, x):
        return x

    def training_step(self, batch, batch_idx):
        self.metric(batch[0])
        return None

    def training_epoch_end(self, training_step_outputs):
        print(f"Computed sum: {self.metric.compute()} GPU {torch.distributed.get_rank()}")
        print(f"Actual sum: {data.sum()} GPU {torch.distributed.get_rank()}")

    def train_dataloader(self):
        return torch.utils.data.DataLoader(
            dataset, batch_size=1, sampler=torch.utils.data.distributed.DistributedSampler(dataset, shuffle=False))

    def configure_optimizers(self):
        return None


pl.Trainer(gpus=2, max_epochs=1,
           distributed_backend='ddp', progress_bar_refresh_rate=0,
           replace_sampler_ddp=False).fit(IdentityModel())
```

The code above creates a dataset with {0,...,9} and creates a metric that is supposed to sum up all elements, which yields a total sum of 45 per epoch. The computation is performed with DDP and 2 gpus. Despite the expected output to be 45, the code return 41. The reason for this is the issue mentioned above. Here is the simulation of what happens. I provide pairs (a,b) indicating that a is the metric state of GPU 0 and b the metric state of GPU 1.

```
(0,1)  ->  after batch 0 containing (0,1)
(0,0)  ->  DDP buffer sync
(2,3)  ->  after batch 1 containing (2,3)
(2,2)  ->  DDP buffer sync
(6,7)  ->  after batch 2 containing (4,5)
(6,6)  ->  DDP buffer sync
(12,13)  ->  after batch 2 containing (6,7)
(12,12)  ->  DDP buffer sync
(20,21)  ->  after batch 2 containing (8,9)
41 -> final metric synchronization 
```

Let me know if I should open a separate issue.

**Comment by SkafteNicki** [comment]
@rotabulo I see the problem now. I am really not sure why this is not captured by our tests.
It seems that the solution to this is not using `register_buffer`, as it is not only making troubles in `ddp` but also `dp`. Instead we should probably settle for just allocating our states as normal tensors, and then make sure our self that they get moved to the correct device (which is the clear benefit of using `register_buffer`).

**Comment by rotabulo** [comment]
@SkafteNicki this is also my fallback solution. The only think is that you have to take care of saving and loading the metrics manually.

**Comment by SkafteNicki** [comment]
@rotabulo by this you mean that they should manually be added to the modules `state_dict` right?
Would you be up for trying to fix this? Or if you don`t have time, could you open a PR adding the test you wrote so we have a starting point for fixing this.

**Comment by rotabulo** [comment]
@SkafteNicki unfortunately I don't have time to fix or make a PR. To be honest, I find this synchronization done by DDP nosense and I would rather fix that, considering that Lightning already introduces some modifications on top of DDP. Shall I open a separate issue, or is what I wrote here enough?

**Comment by SkafteNicki** [comment]
@rotabulo the details you have provided here are fine :]
Thank you for identifying this pitfall in the metrics package, I will try to come up with a solution.
